README.md
# next-select

A vanilla JS customisable select/dropdown box plugin

## Support

![](/public/next-select.gif)

## Features

- No Dependencies
- JS: 29kb - 5kb gzip
- CSS: 7kb - 1kb gzip
- Single Select
- Multi Select
- User Addable Options
- Html Options
- Settable Data
- Callback Events
- Placeholders
- Search
- Disable Options
- Light CSS
- Light Color Scheme
- Style and Class Inheritance
- Clean Animations
- Performant
- Typescript


## Installation

```bash
npm install next-select
```

### or

```html
<script src="https://unpkg.com/next-select@latest/dist/next-select.min.js"></script>
<link rel="stylesheet" href="https://unpkg.com/next-select@latest/dist/select.css" />
```

## Simple Usage

```javascript
import NextSelect from 'next-select'

new NextSelect({
  select: '#selectElement',
})
```

```html
<select id="selectElement">
  <option value="value1">Value 1</option>
</select>
```

## Data

Data is an array of objects that represent both option and optgroups.

See below for list of data types

```javascript
new NextSelect({
  select: '#selectElement',

  // Array of Option objects
  data: [{ text: 'Value 1', value: 'value1' }],

  // or

  // Array of Optgroups and/or Options
  data: [{ label: 'Optgroup Label', options: { text: 'Value 1', value: 'value1' } }],
})
```

## Data Types

```javascript
// <optgroup>
var optgroup = {
  label: 'label', // Required
  selectAll: false, // Optional - default false
  closable: 'off', // Optional - default 'off' - 'off', 'open', 'close'
  options: [], // Required - value is an array of options
}

// <option>
var option = {
  text: 'text', // Required
  value: 'value', // Optional - value will be set by text if not set
  html: '<b>Html</b>', // Optional - if set, used for display purposes
  selected: false, // Optional - default is false
  display: true, // Optional - default is true
  disabled: false, // Optional - default is false
  mandatory: false, // Optional - default is false
  placeholder: false, // Optional - default is false
  class: '', // Optional - default is not set
  style: '', // Optional - default is not set
  data: {}, // Optional - If you have data attributes
}
```

## Settings

Settings are a list of fields that help customize how NextSelect operates

```javascript
new NextSelect({
  select: '#selectElement',

  settings: {
    // Below are a list of optional fields
    // their values are the defaults
    disabled: false,
    alwaysOpen: false,
    showSearch: true,
    searchPlaceholder: 'Search',
    searchText: 'No Results',
    searchingText: 'Searching...',
    searchHighlight: false,
    closeOnSelect: true,
    contentLocation: document.body,
    contentPosition: 'absolute',
    openPosition: 'auto', // options: auto, up, down
    placeholderText: 'Select Value',
    allowDeselect: false,
    hideSelected: false,
    showOptionTooltips: false,
    minSelected: 0,
    maxSelected: 1000,
    timeoutDelay: 200,
    maxValuesShown: 20,
    maxValuesMessage: '{number} selected',
  },
})
```

## Events

Events are function callbacks for when certain actions happen

```javascript
new NextSelect({
  select: '#selectElement',

  events: {
    search: (searchValue: string, currentData: DataArray) => Promise<DataArrayPartial> | DataArrayPartial
    searchFilter: (option: Option, search: string) => boolean
    addable: (value: string) => Promise<OptionOptional | string> | OptionOptional | string
    beforeChange: (newVal: Option[], oldVal: Option[]) => boolean | void
    afterChange: (newVal: Option[]) => void
    beforeOpen: () => void
    afterOpen: () => void
    beforeClose: () => void
    afterClose: () => void
    error: (err: Error) => void
  },
})
```


CHANGELOG.md
# Change Log

All notable changes to the "next-select" project will be documented in this file.

## [0.0.0] - 2023-09-01

- Initial release


src/settings.ts
import { generateID } from "./helpers";

export type SettingsPartial = Partial<Settings>;

export default class Settings {
  public id: string = ""; // Primary ID for the select
  public style: string = ""; // Style attribute from the select element
  public class: string[] = []; // Class attribute from the select element

  // Dynamic settings
  public isMultiple: boolean = false;
  public isOpen: boolean = false;
  public isFullOpen: boolean = false;
  public intervalMove: NodeJS.Timeout | null = null;

  // Fields set from constructor
  public disabled: boolean;
  public alwaysOpen: boolean;
  public showSearch: boolean;
  public searchPlaceholder: string;
  public searchText: string;
  public searchingText: string;
  public searchHighlight: boolean;
  public closeOnSelect: boolean;
  public contentLocation: HTMLElement;
  public contentPosition: "relative" | "absolute";
  public openPosition: "auto" | "up" | "down";
  public placeholderText: string;
  public allowDeselect: boolean;
  public hideSelected: boolean;
  public showOptionTooltips: boolean;
  public minSelected: number;
  public maxSelected: number;
  public timeoutDelay: number;
  public maxValuesShown: number;
  public maxValuesMessage: string;

  constructor(settings?: SettingsPartial) {
    if (!settings) {
      settings = {};
    }

    this.id = "next-select-" + generateID();
    this.style = settings.style || "";
    this.class = settings.class || [];

    this.disabled = settings.disabled !== undefined ? settings.disabled : false;
    this.alwaysOpen = settings.alwaysOpen !== undefined ? settings.alwaysOpen : false;
    this.showSearch = settings.showSearch !== undefined ? settings.showSearch : true;
    this.searchPlaceholder = settings.searchPlaceholder || "Search";
    this.searchText = settings.searchText || "No Results";
    this.searchingText = settings.searchingText || "Searching...";
    this.searchHighlight =
      settings.searchHighlight !== undefined ? settings.searchHighlight : false;
    this.closeOnSelect = settings.closeOnSelect !== undefined ? settings.closeOnSelect : true;
    this.contentLocation = settings.contentLocation || document.body;
    this.contentPosition = settings.contentPosition || "absolute";
    this.openPosition = settings.openPosition || "auto"; // options: auto, up, down
    this.placeholderText =
      settings.placeholderText !== undefined ? settings.placeholderText : "Select Value";
    this.allowDeselect = settings.allowDeselect !== undefined ? settings.allowDeselect : false;
    this.hideSelected = settings.hideSelected !== undefined ? settings.hideSelected : false;
    this.showOptionTooltips =
      settings.showOptionTooltips !== undefined ? settings.showOptionTooltips : false;
    this.minSelected = settings.minSelected || 0;
    this.maxSelected = settings.maxSelected || 1000;
    this.timeoutDelay = settings.timeoutDelay || 200;
    this.maxValuesShown = settings.maxValuesShown || 20;
    this.maxValuesMessage = settings.maxValuesMessage || "{number} selected";
  }
}
